contrib/win32/win32compat/w32-sshfileperm.c

```diff
@@ -0,0 +1,342 @@
+		FreeSid(user_sid);
+	if(name_utf16)
+		free(name_utf16);
+	return ret;
+}
+
+static BOOL
+is_sshd_account(PSID user_sid) {	
+	wchar_t user_name[UNCLEN], full_name[UNCLEN + DNLEN + 2];
+	DWORD name_length = UNCLEN, domain_name_length = 0, full_name_len = UNCLEN + DNLEN + 2;
+	SID_NAME_USE sid_type = SidTypeInvalid;
+	BOOL ret = FALSE;
+
+	if (LookupAccountSidLocalW(user_sid, user_name, &name_length, full_name, &full_name_len, &sid_type) == FALSE)
+	{
+		debug3("LookupAccountSidLocalW() failed with error: %d. ", GetLastError());
+		errno = ENOENT;
+		return FALSE;
+	}
+	domain_name_length = wcslen(full_name);
+	full_name[domain_name_length] = L'\\';
+	wmemcpy(full_name + domain_name_length + 1, user_name, wcslen(user_name)+1);
+	return (wcsicmp(full_name, SSHD_ACCOUNT) == 0);
+}
+
+/*
+ * Check if the user is in local administrators group 
+ * currently only check if the user is directly in the group
+ * Returns TRUE if the user is in administrators group; otherwise false; 
+*/
+static BOOL
+is_admin_account(PSID user_sid)
+{
+	DWORD entries_read = 0, total_entries = 0, i = 0, name_length = UNCLEN, domain_name_length = DNLEN;
+	LPLOCALGROUP_MEMBERS_INFO_1 local_groups_member_info = NULL;
+	PSID admins_sid = NULL;
+	wchar_t admins_group_name[UNCLEN], domain_name[DNLEN];
+	SID_NAME_USE sid_type = SidTypeInvalid;
+	NET_API_STATUS status;
+	BOOL ret = FALSE;	
+
+	if (ConvertStringSidToSidW(L"S-1-5-32-544", &admins_sid) == FALSE ||
+		(IsValidSid(user_sid) == FALSE)) {
+		debug3("ConvertStringSidToSidW failed with error code: %d.", GetLastError());
+		goto done;
+	}
+
+	if (LookupAccountSidLocalW(admins_sid, admins_group_name, &name_length,
+		domain_name, &domain_name_length, &sid_type) == FALSE) {
+		debug3("LookupAccountSidLocalW() failed with error: %d. ", GetLastError());
+		errno = ENOENT;
+		goto done;
+	}
+	
+	status = NetLocalGroupGetMembers(NULL, admins_group_name, 1, (LPBYTE*)&local_groups_member_info,
+	MAX_PREFERRED_LENGTH, &entries_read, &total_entries, NULL);
+	if (status != NERR_Success) {
+		debug3("NetLocalGroupGetMembers failed with error code: %d.", status);
+		goto done;
+	}
+
+	for (i = 0; i < entries_read; i++) {
+		if (local_groups_member_info[i].lgrmi1_sidusage == SidTypeDeletedAccount)
+			continue;		
+		else if(EqualSid(local_groups_member_info[i].lgrmi1_sid, user_sid)) {
+			ret = TRUE;
+			break;
+		}
+	}
+
+done:	
+	if (local_groups_member_info)
+		NetApiBufferFree(local_groups_member_info);
+	if(admins_sid)
+		LocalFree(admins_sid);
+	return ret;
+}
+
+int
+set_secure_file_permission(const char *name, struct passwd * pw)
+{
+	PSECURITY_DESCRIPTOR pSD = NULL;
+	PSID owner_sid = NULL;
+	PACL dacl = NULL;
+	wchar_t *name_utf16 = NULL, *sid_utf16 = NULL, sddl[256];
+	DWORD error_code = ERROR_SUCCESS;
+	struct passwd * pwd = pw;
+	BOOL present, defaulted;
+	int ret = 0;
+
+	if (pwd == NULL)
+		if ((pwd = getpwuid(0)) == NULL)
+			fatal("getpwuid failed.");
+
+	if (ConvertStringSidToSid(pwd->pw_sid, &owner_sid) == FALSE) {
```

Comment:
but the function you mentioned is not needed here. Here is to convert the urf8 string to a sid.